packages/@react-facet/core/src/hooks/useFacetCallback.ts

```diff
@@ -1,47 +1,158 @@
+  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5, v6: V6, v7: V7, v8: V8) => C,
+  dependencies: unknown[],
+  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>, Facet<V4>, Facet<V5>, Facet<V6>, Facet<V7>, Facet<V8>],
+): C
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function useFacetCallback<M, V, V1, V2, V3, V4, V5, V6, V7, V8, V9, C extends (...args: any[]) => M | NoValue>(
+  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5, v6: V6, v7: V7, v8: V8, v9: V9) => C,
+  dependencies: unknown[],
+  facet: [Facet<V>, Facet<V1>, Facet<V2>, Facet<V3>, Facet<V4>, Facet<V5>, Facet<V6>, Facet<V7>, Facet<V8>, Facet<V9>],
+): C
+
+export function useFacetCallback<
+  M,
+  V,
+  V1,
+  V2,
+  V3,
+  V4,
+  V5,
+  V6,
+  V7,
+  V8,
+  V9,
+  V10,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  C extends (...args: any[]) => M | NoValue,
+>(
+  callback: (v: V, v1: V1, v2: V2, v3: V3, v4: V4, v5: V5, v6: V6, v7: V7, v8: V8, v9: V9, v10: V10) => C,
+  dependencies: unknown[],
+  facet: [
+    Facet<V>,
+    Facet<V1>,
+    Facet<V2>,
+    Facet<V3>,
+    Facet<V4>,
+    Facet<V5>,
+    Facet<V6>,
+    Facet<V7>,
+    Facet<V8>,
+    Facet<V9>,
+    Facet<V10>,
+  ],
+): C
 
 /**
  * Creates a callback that depends on the value of a facet.
  * Very similar to `useCallback` from `React`
  *
- * @param callback function callback that receives the current facet value and the arguments passed to the callback
- * @param dependencies variable used by the map that are available in scope (similar as dependencies of useCallback)
- * @param facet facet that the callback listens to
+ * @param callback function callback that receives the current facet values and the arguments passed to the callback
+ * @param dependencies variable used by the callback that are available in scope (similar as dependencies of useCallback)
+ * @param facets facets that the callback listens to
  *
  * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
  * Having this as the second argument allows the linter to work.
  */
-export function useFacetCallback<V, T>(
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  callback: (value: V) => (...args: any[]) => T | NoValue,
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  dependencies: any[],
-  facet: Facet<V>,
-) {
-  const facetRef = useRef<Option<V>>(facet.get())
-
-  useFacetEffect(
-    (value) => {
-      facetRef.current = value
-    },
-    [],
-    facet,
-  )
+export function useFacetCallback<M, C extends (...args: unknown[]) => M | NoValue>(
+  callback: (...args: unknown[]) => C,
+  dependencies: unknown[],
+  facets: Facet<unknown>[],
+): C {
+  const facetsRef = useRef<Option<unknown>[]>(facets.map(() => NO_VALUE))
+
+  useEffect(() => {
+    const unsubscribe = facets.map((facet, index) =>
+      facet.observe((value) => {
```

Comment:
not super obvious that it returns the unsubscribe function into the map, maybe worth adding curly braces and doing explicit return? in case someone wants to add something before this line in the future, so we don't miss unsubscribe

on the other hand we will "undefined is not a function" so maybe it's fine to leave it as is

Reply:
Done at e9e1d62